runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,227 @@
+        }
+        if (subscriptions.containsKey(streamListener)) {
+            throw new SubscriberAlreadyExistsException(
+                    "StreamManager:subscriber already registered "
+                            + streamListener + ". Maybe call unregister first?");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}",
+                streamListener, startAddress, namespace, tablesOfInterest.toString());
+        SubscriptionThread st = new SubscriptionThread(this, streamListener,
+                namespace, tablesOfInterest, startAddress);
+        subscriptions.put(streamListener, st);
+        st.start();
+    }
+
+    /**
+     * Unsubscribe a prior subscription.
+     *
+     * @param streamListener Client listener.
+     */
+    public synchronized void unsubscribe(@Nonnull StreamListener streamListener) {
+        SubscriptionThread st = subscriptions.remove(streamListener);
+        if (st == null) {
+            log.warn("StreamManager::unsubscribe has no context for {}", streamListener.toString());
+            return;
+        }
+        st.shutdown.set(true);
+        st.interrupt();
+        log.info("Unsubscribed StreamListener {}", streamListener.toString());
+    }
+
+    class SubscriptionThread extends Thread {
+        /**
+         * The callback context
+         */
+        private final StreamListener listener;
+
+        /**
+         * Parent class to manage errors.
+         */
+        @Getter
+        private final StreamManager streamManager;
+
+        /**
+         * Namespace of interest.
+         */
+        @Getter
+        private final String namespace;
+
+        /**
+         * Tables of interest.
+         *
+         * Map of (Stream Id - TableSchema)
+         */
+        @Getter
+        private final Map<UUID,
+                TableSchema<? extends Message, ? extends Message, ? extends Message>> tablesOfInterest;
+
+        /**
+         * Starting address.
+         */
+        private final long startAddress;
+
+        public AtomicBoolean shutdown;
+
+        /**
+         * A reference to the underlying stream
+         */
+        @Getter
+        private final IStreamView txnStream;
+
+        public void run() {
+            try {
+                log.info("{}:Seeking txStream to {}", listener.toString(), startAddress + 1);
+                txnStream.seek(startAddress + 1);
+            } catch (Throwable throwable) {
+                log.warn("{}:onError at seek to {}", listener.toString(), startAddress + 1);
+                listener.onError(throwable);
+                streamManager.unsubscribe(listener);
+                return;
+            }
+
+            try {
+                while (!shutdown.get()) {
+                    while (!txnStream.hasNext()) {
+                        log.trace("{}::run() all caught up. Taking a 50ms nap", listener.toString());
+                        int DEFAULT_TIMEOUT_IN_MILLI = 50;
```

Comment:
please rename the variable, it is not a constant

Reply:
done. thanks

Reply:
done. thanks